// TODO(amwolff): Move things here and there according to the linter rules.
syntax = "proto3";

package matmult;

service Performer {
  rpc MultiplyMatrices(PerformerRequest) returns (PerformerResponse) {}
}

message Matrix {
  // Eigen uses column-major order to store its matrices:
  // https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html
  // I think it's fair to adjust to Eigen.
  message Column {
    repeated double coefficients = 1 [packed=true];
  }
  repeated Column columns = 1;
}

message PerformerRequest {
  Matrix multiplier = 1;
  Matrix multiplicand = 2;
}

message PerformerResponse {
  Matrix result = 1;
}
